### ISSUE TITLE
`lint.py_run` does not use `setup.cfg`

### ISSUE THREAD

**Issue by bersbersbers** [PyCQA/pylint#5703]
### Bug description

This is `bug.py`:

```python
import subprocess
import tempfile

from pylint import epylint as lint

with tempfile.NamedTemporaryFile(mode="w") as temp:
    temp.write('"""Should issue useless-suppression."""\n')
    temp.write('print("Hello")  # pylint:disable=no-member\n')
    temp.flush()

    print("Does not issue useless-suppression => does not use setup.cfg")
    (stdout, _) = lint.py_run(temp.name, return_std=True)
    print(stdout.getvalue())

    print("Does issue useless-suppression => does use setup.cfg")
    proc = subprocess.run(["pylint", temp.name], capture_output=True, text=True)
    print(proc.stdout)

```

### Configuration

This is `setup.cfg`:

```ini
[pylint.'MESSAGES CONTROL']
enable = useless-suppression
```

### Command used

```shell
python bug.py
```

### Pylint output

```shell
$ python bug.py
Does not issue useless-suppression => does not use setup.cfg

 ------------------------------------
 Your code has been rated at 10.00/10
 
 
Does issue useless-suppression => does use setup.cfg
************* Module tmpl1ha377e
/tmp/tmpl1ha377e:2:0: I0021: Useless suppression of 'no-member' (useless-suppression)

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

### Expected behavior

Should issue `useless-suppression` in both cases.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.9.10 (main, Jan 17 2022, 14:25:44) 
[GCC 7.5.0]
```

### OS / Environment

OpenSUSE Leap 15.2 Linux

**Comment by bersbersbers** [comment]
Here's an extended example which shows that `pylint.epylint.py_run`, `pylint.lint.Run`, and `subprocess.call(["pylint", ...])` all three behave differently:

```python
"""Demonstrate bug."""
import subprocess
import tempfile

import pylint.epylint
import pylint.lint

with tempfile.NamedTemporaryFile(mode="w") as temp:
    temp.write('"""Should issue useless-suppression."""\n')
    temp.write('print("Hello")  # pylint:disable=no-member\n')
    temp.flush()

    print("***** Does not issue useless-suppression => does not use setup.cfg")
    print("---------------------------------------------------------------------------")
    (stdout, _) = pylint.epylint.py_run(temp.name, return_std=True)
    print(stdout.getvalue())
    print()

    print("***** Does not issue useless-suppression => does not use setup.cfg")
    print("***** In addition, issues a subprocess-run-check")
    print("---------------------------------------------------------------------------")
    pylint.lint.Run([__file__], exit=False)
    print()
    print()

    print("***** Does issue useless-suppression => does use setup.cfg")
    print("---------------------------------------------------------------------------")
    proc = subprocess.run(["pylint", temp.name], capture_output=True, text=True)
    print(proc.stdout)
```

Output:
```
***** Does not issue useless-suppression => does not use setup.cfg
---------------------------------------------------------------------------

 ------------------------------------
 Your code has been rated at 10.00/10
 
 

***** Does not issue useless-suppression => does not use setup.cfg
***** In addition, issues a subprocess-run-check
---------------------------------------------------------------------------
************* Module xxx.bug
xxx/bug.py:28:11: W1510: Using subprocess.run without explicitly set `check` is not recommended. (subprocess-run-check)

------------------------------------------------------------------
Your code has been rated at 9.57/10 (previous run: 9.57/10, +0.00)



***** Does issue useless-suppression => does use setup.cfg
---------------------------------------------------------------------------
************* Module tmpes14fa1h
/tmp/tmpes14fa1h:2:0: I0021: Useless suppression of 'no-member' (useless-suppression)

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

**Comment by DanielNoord** [comment]
@bersbersbers This seems to be related to exploring configuration files (obviously).

We're in the middle of migrating our options parsing from `optparse` to `argparse` which will probably also touch upon this.
However, this will probably take some more time, as we need `2.13` released before we can continue working on it.

I'm not sure what your specific use case is, but I think it might take some time before we get to this issue ourselves. If the matter is more time-sensitive for you I'd happily review a PR or help towards identifying the issue. Grepping for `setup.cfg` should at least show which part of the code is not being called or called incorrectly in those two cases.

**Comment by DanielNoord** [comment]
@bersbersbers Ah, okay! Well `subprocess` seems the way to go indeed.

Could you do something for me though? Could you apply the following patch:
```diff

diff --git a/pylint/config/option_manager_mixin.py b/pylint/config/option_manager_mixin.py
index 67740d334..24122860d 100644
--- a/pylint/config/option_manager_mixin.py
+++ b/pylint/config/option_manager_mixin.py
@@ -60,6 +60,7 @@ class OptionsManagerMixIn:
     """Handle configuration from both a configuration file and command line options"""
 
     def __init__(self, usage, config_file=None):
+       print(config_file)
         self.config_file = config_file
         self.reset_parsers(usage)
         # list of registered options providers
```
And see what gets printed as `config_file`? For my own environment I got quite surprising results.. It found: ` /Users/daniel/.pylintrc`, which I totally forgot about and did not expect to be found.

**Comment by bersbersbers** [comment]
Using this patch and the code from https://github.com/PyCQA/pylint/issues/5703#issue-1107909741, this is what I get:

```
Does not issue useless-suppression => does not use setup.cfg
None
 
 ------------------------------------
 Your code has been rated at 10.00/10
 
 
Does issue useless-suppression => does use setup.cfg
/home/bers/xxx/setup.cfg
************* Module tmpvlpbp99e
/tmp/tmpvlpbp99e:2:0: I0021: Useless suppression of 'no-member' (useless-suppression)

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

**Comment by DanielNoord** [comment]
I found the issue:
https://github.com/PyCQA/pylint/blob/2ab7d536159f5e54c195db7a42f00261c3b67827/pylint/epylint.py#L95-L119

Because you're creating a `tempfile` the `child_path` becomes the temporary file directory, which does not contain `setup.cfg`. I think if you create a `setup.cfg` in the temporary dir this should still work.
Why does the other command behave differently than `subprocess`? Because `pylint` doesn't invoke `epylint` but `pylint.run_pylint()`. The following code does work:
```python
import subprocess
import tempfile

import pylint

with tempfile.NamedTemporaryFile(mode="w") as temp:
    temp.write('"""Should issue useless-suppression."""\n')
    temp.write('print("Hello")  # pylint:disable=no-member\n')
    temp.flush()

    print("Does not issue useless-suppression => does not use setup.cfg")
    try:
        (stdout, _) = pylint.run_pylint([temp.name])
    except SystemExit:
        pass

    print("Does issue useless-suppression => does use setup.cfg")
    proc = subprocess.run(["pylint", temp.name], capture_output=True, text=True)
    print(proc.stdout)
```

`run_pylint` prints to `std_out` automatically so that's no longer needed.

**Comment by bersbersbers** [comment]
Thanks! `pylint.run_pylint` does not take any arguments, but `pylint.lint.Run` does. This works:

```python
import subprocess
import tempfile

import pylint.lint

with tempfile.NamedTemporaryFile(mode="w") as temp:
    temp.write('"""Should issue useless-suppression."""\n')
    temp.write('print("Hello")  # pylint:disable=no-member\n')
    temp.flush()

    print("Does issue useless-suppression => does use setup.cfg")
    pylint.lint.Run([temp.name], exit=False)

    print("Does issue useless-suppression => does use setup.cfg")
    proc = subprocess.run(["pylint", temp.name], capture_output=True, text=True)
    print(proc.stdout)
```

I would have noticed that this one works if my example in https://github.com/PyCQA/pylint/issues/5703#issuecomment-1016482552 had not had a copy-and-past error...

So this working for me. Just for my understanding - there is still a bug to be fixed in `epylint`, right? Because according to the docs, `pylint` does (also) look for a `setup.cfg` in the working directory.

**Comment by DanielNoord** [comment]
Yeah calling `Run` directly also works!

Hm, I'm not sure about `epylint`. I believe it is the `pylint` used for emacs and I think there is a reason why they do that path resolving before instantiating `Run`. I don't think we really have any issues reported about emacs users not being able to configure their `pylint`. I'm going to wait on @Pierre-Sassoulas but I think we can close this.

_There is an issue that epylint and tempfile don't really work together but I don't think that's something that we want to support. Happy to review a PR but I think any of current contributors will much rather fix any of the 700+ open issues._

**Comment by Pierre-Sassoulas** [comment]
(As far as I know epylint has no users because I never seen any open an issue or a merge request. I'm not using emacs myself, so epylint bugs are probably going to take forever to fix.)

If you think the documentation could be made clearer/better with the knowledge you just acquired feel free to open a documentation merge request @bersbersbers,, otherwise I think we can close :)

**Comment by DanielNoord** [comment]
Closing this as I think all issues have been resolved.
